### PR TITLE
storage/engine: async engine api

### PR DESCRIPTION
**Pull request by disksing**
For https://github.com/pingcap/tikv/issues/807

@zhangjinpeng1987 @BusyJay @ngaut @siddontang

### CODE REVIEW

File: src/storage/engine/rocksdb.rs [hunk 1]
```diff
@@ -12,20 +12,25 @@
...
 
 
 pub struct EngineRocksdb {
-    db: DB,
+    db: Arc<DB>,
     // only use for memory mode
     temp_dir: Option<TempDir>,
+    // ThreadPool is handy for executing task asynchronously, each Engine will
+    // contain only 1 thread.
+    pool: Mutex<ThreadPool>,
```
Comment (siddontang) [inline src/storage/engine/rocksdb.rs]: why not using util::worker directly?

File: src/storage/engine/rocksdb.rs [hunk 2]
```diff
@@ -39,8 +44,9 @@ impl EngineRocksdb {
             _ => (path.to_owned(), None),
         };
         Ok(EngineRocksdb {
-            db: try!(rocksdb::new_engine(&path, cfs)),
+            db: Arc::new(try!(rocksdb::new_engine(&path, cfs))),
```
Comment (siddontang) [inline src/storage/engine/rocksdb.rs]: I think new_engine can return Arc now.

**Comment by siddontang** [comment]
Rest LGTM

File: src/storage/engine/rocksdb.rs (hunk 1 above)
Comment (disksing) [inline src/storage/engine/rocksdb.rs]: To use `util::worker`, we typically will have to define a enum Msg, then implement a `Runner` to handle different types of Msgs.
While `ThreadPool` has a `fn execute<F>(&self, job: F) where F: FnOnce() + Send + 'static`, which is more easy to use.

File: src/storage/engine/rocksdb.rs (hunk 2 above)
Comment (disksing) [inline src/storage/engine/rocksdb.rs]: I prefer not to do it. There are still cases we create a DB but do not need it to be shared between threads. For example, `tikv-dump`.

Comment (ngaut) [inline src/storage/engine/rocksdb.rs]: +1

File: src/storage/engine/rocksdb.rs (hunk 1 above)
Comment (BusyJay) [inline src/storage/engine/rocksdb.rs]: I think the `worker` way is more clear. And you can hold the context only in runner without cloning it every time.

**Comment by disksing** [comment]
PTAL @siddontang @BusyJay 
I have also replaced `Engine::close()` with `Drop`, please help to review if it's okay. And I'm not quite sure if the CI failure https://travis-ci.org/pingcap/tikv/jobs/144667055 is related to the last commit.

File: src/storage/engine/rocksdb.rs [hunk 3]
```diff
@@ -11,26 +11,53 @@
...
+impl Runnable<Task> for Runner {
+    fn run(&mut self, t: Task) {
+        match t {
+            Task::Write(modifies, cb) => cb(write_modifies(&self.0, modifies)),
+            Task::Snapshot(cb) => cb(Ok(box RocksSnapshot::new(self.0.clone()))),
+        }
+    }
+}
+
+impl Display for Runner {
```
Comment (BusyJay) [inline src/storage/engine/rocksdb.rs]: Why impl Display for runner?

**Comment by BusyJay** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**